compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -1947,17 +1948,20 @@ boolean checkTypeAndVarCountConsistency(BLangTupleVariable varNode, BTupleType t
                 case TypeTags.UNION:
                     Set<BType> unionType = types.expandAndGetMemberTypesRecursive(varNode.type);
                     List<BType> possibleTypes = new ArrayList<>();
-                    for (BType type :unionType) {
-                        if (!(TypeTags.TUPLE == type.tag &&
-                                (varNode.memberVariables.size() == ((BTupleType) type).tupleTypes.size())) &&
-                                TypeTags.ANY != type.tag &&
-                                TypeTags.ANYDATA != type.tag) {
+                    for (BType type : unionType) {
+                        if (TypeTags.TUPLE == type.tag &&
+                                !checkMemVarCountMatchWithMemTypeCount(varNode, (BTupleType) type)) {
+                            dlog.error(varNode.pos, DiagnosticErrorCode.INVALID_LIST_BINDING_PATTERN);
+                            return false;
```

Comment:
Is it correct to log the error here? What if there is another type in the union that matches?